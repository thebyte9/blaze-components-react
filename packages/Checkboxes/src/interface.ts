import { FunctionComponent } from "react";

interface IErrorMessage {
  message: string | JSX.Element;
  icon?: string;
}

interface ICheckBoxesProps {
  options?: any[] | object;
  returnBoolean?: boolean;
  onChange: ({
    event,
    value,
    data,
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    value: boolean | object;
    data: object[];
  }) => void;
  error?: boolean;
  name?: string;
  validationMessage: string | JSX.Element;
  utils: {
    uniqueId: (element: any) => string;
    classNames: (className: string | object, classNames?: object) => string;
    ErrorMessage: FunctionComponent<IErrorMessage>;
  };
}

export { IErrorMessage, ICheckBoxesProps };
