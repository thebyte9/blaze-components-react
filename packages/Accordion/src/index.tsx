import React, { useEffect, useState } from "react";
import AccordionContent from "./AccordionContent";
import AccordionContentDetails from "./AccordionContent/AccordionContentDetails";
import AccordionContentFooter from "./AccordionContent/AccordionContentFooter";
import AccordionHeader from "./AccordionHeader";

interface IAccordionProps {
  children: [JSX.Element, JSX.Element];
  open?: boolean;
}

const Accordion = ({ children, open }: IAccordionProps): JSX.Element => {
  const flex = "flex";
  const none = "none";

  const [accordionStatus, setAccordionStatus] = useState<string>(
    open ? flex : none
  );

  const [header, content]: [JSX.Element, JSX.Element] = children;

  useEffect(() => {
    setAccordionStatus(open ? flex : none);
  }, [open]);

  const isActive: boolean = accordionStatus === flex;
  const toggleAccordion = (): void =>
    setAccordionStatus(isActive ? none : flex);

  return (
    <div className="accordion">
      <div className="accordion__header">
        {header}
        <div
          className="icon-button icon-button--accordion"
          onClick={toggleAccordion}
        >
          <i className="material-icons">keyboard_arrow_down</i>
        </div>
      </div>
      <div
        className="accordion__content-wrapper"
        style={{ display: accordionStatus }}
      >
        {content}
      </div>
    </div>
  );
};

Accordion.Content = AccordionContent;
Accordion.ContentDetails = AccordionContentDetails;
Accordion.ContentFooter = AccordionContentFooter;
Accordion.Header = AccordionHeader;
export default Accordion;
