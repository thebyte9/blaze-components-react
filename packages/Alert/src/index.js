import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import Button from '@blaze-react/button';

const Alert = ({ children, close, icon, type, ...attrs }) => {
  const [offModal, setModalOff] = useState(false);
  const assignType = type && `alert--${type}`;
  const isDismissable = close && 'alert--dismissable';
  const withIcon = icon && 'alert--icon';

  const renderAlert = (
    <div className={`alert ${assignType} ${isDismissable} ${withIcon}`} {...attrs}>
      {icon && <i className="material-icons">{icon}</i>}
      {children}
      {close && (
        <Button onClick={() => setModalOff(true)} className="icon-button icon-button--close">
          <i className="material-icons">close</i>
        </Button>
      )}
    </div>
  );

  return <Fragment>{!offModal && renderAlert}</Fragment>;
};

Alert.propTypes = {
  icon: PropTypes.string,
  type: PropTypes.string,
  close: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
};

Alert.defaultProps = {
  icon: '',
  type: '',
  close: false,
  children: 'No content'
};

export default Alert;
