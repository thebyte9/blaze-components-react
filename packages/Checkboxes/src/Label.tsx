import React from "react";

const parsedLabel = ({
  defaultId,
  label
}: {
  defaultId: any;
  label?: string | [string, string];
}) => {
  if (Array.isArray(label)) {
    const [labelText, labelLongerText] = label;
    return (
      <label htmlFor={defaultId}>
        <span>{labelText}</span>
        <span>{labelLongerText}</span>
      </label>
    );
  }
  return (
    <label htmlFor={defaultId}>
      <span>{label}</span>
    </label>
  );
};

export default parsedLabel;