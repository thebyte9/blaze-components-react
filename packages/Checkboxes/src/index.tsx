import withUtils from "@blaze-react/utils";
import React, { Fragment, FunctionComponent, useEffect, useState } from "react";
interface IErrorMessage {
  message: string | JSX.Element;
  icon?: string;
}

interface ICheckboxesProps {
  options?: any[] | object;
  returnBoolean?: boolean;
  onChange: ({
    event,
    value,
    data
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    value: boolean | object;
    data: object[];
  }) => void;
  error?: boolean;
  validationMessage: string | JSX.Element;
  utils: {
    uniqueId: (element: any) => string;
    classNames: (className: string | object, classNames?: object) => string;
    ErrorMessage: FunctionComponent<IErrorMessage>;
  };
}
const Checkboxes: FunctionComponent<ICheckboxesProps> = ({
  returnBoolean,
  onChange,
  options,
  error,
  validationMessage,
  utils: { uniqueId, classNames, ErrorMessage },
  ...attrs
}): JSX.Element => {
  const formatedOptions = Array.isArray(options) ? options : [options];

  const [data, setData]: any = useState(formatedOptions);

  useEffect(() => setData(formatedOptions), [options]);

  const toggle = ({
    event,
    item,
    key
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    item: any;
    key: number;
  }): void => {
    if (item.disabled) {
      return;
    }

    data[key].checked = !item.checked;
    setData([...data]);

    let value = data.filter(
      ({ checked }: { checked: boolean }): boolean => checked
    );

    if (returnBoolean) {
      value = !!value.length;
    }

    onChange({ event, value, data });
  };

  return (
    <Fragment>
      {data.map(
        (item: any, key: number): JSX.Element => {
          const {
            checked = false,
            value,
            disabled,
            required,
            label,
            show = true,
            id = uniqueId(item)
          } = item;

          if (!show) {
            return <Fragment key={id} />;
          }

          const checkboxClassName = classNames(
            "form-field form-field--checkbox",
            { required }
          );

          return (
            <div
              data-testid={`checkbox-${key + 1}`}
              key={id}
              className={checkboxClassName}
              onClick={(event): void => toggle({ event, item, key })}
              role="button"
            >
              <input
                readOnly
                type="checkbox"
                className="form-checkbox"
                value={value}
                disabled={disabled}
                checked={checked}
                required={required}
                id={id}
                {...attrs}
              />
              <label htmlFor={id}>{label}</label>
            </div>
          );
        }
      )}
      {error && <ErrorMessage message={validationMessage} />}
    </Fragment>
  );
};
Checkboxes.defaultProps = {
  error: false,
  options: [],
  returnBoolean: false,
  validationMessage: "This field is required"
};
export default withUtils(Checkboxes);
