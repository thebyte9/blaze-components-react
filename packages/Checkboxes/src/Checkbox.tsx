import withUtils from "@blaze-react/utils";
import React, { useState } from "react";
import Label from "./Label";

const Checkbox = ({
  checked,
  value,
  disabled,
  required,
  label,
  show,
  name,
  id,
  onChange,
  full,
  utils: { uniqueId, classNames },
  ...attrs
}: any) => {
  const [isChecked, setIsChecked] = useState(checked);

  const handleCheckboxChange = (
    event: React.MouseEvent<HTMLDivElement>
  ): void => {
    onChange({
      event,
      value: {
        checked: !isChecked,
        disabled,
        id,
        label,
        name,
        required,
        show,
        value,
      },
    });
    setIsChecked(!isChecked);
  };

  const checkboxClassName = classNames("form-field form-field--checkbox", {
    required,
  });

  const labelClassName = classNames({
    "form-field--checkbox-full": full,
  });

  const defaultId = id || uniqueId(name);
  const inputId = `${defaultId}-checkbox`;
  const wrapperId = `${defaultId}-wrapper`;

  return (
    <div
      key={wrapperId}
      className={checkboxClassName}
      role="button"
      id={wrapperId}
    >
      <input
        data-testid={attrs.testId}
        readOnly
        type="checkbox"
        className="form-checkbox"
        value={value}
        disabled={disabled}
        checked={checked}
        required={required}
        id={inputId}
        name={name}
        onClick={handleCheckboxChange}
        data-cy={attrs["data-cy"]}
      />
      <div onClick={handleCheckboxChange} className={labelClassName}>
        <Label defaultId={inputId} label={label} />
      </div>
    </div>
  );
};

export default withUtils(Checkbox);
