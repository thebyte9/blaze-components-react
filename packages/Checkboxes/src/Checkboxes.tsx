import ErrorMessage from '../../Utils/src/ErrorMessage';
import React, { FunctionComponent, useEffect, useState } from "react";
import Checkbox from "./Checkbox";
import { v4 as uuidv4 } from 'uuid';

interface IErrorMessage {
  message: string | JSX.Element;
  icon?: string;
}

export interface ICheckboxItem {
  checked: boolean;
  id: string | number;
  value: any;
}

interface ICheckBoxesProps {
  options?: any[] | Record<string, unknown>;
  returnBoolean?: boolean;
  onChange: ({
    event,
    value,
    data,
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    value: ICheckboxItem[];
    data: Record<string, unknown>[];
  }) => void;
  error?: boolean;
  name?: string;
  validationMessage?: string | JSX.Element;
  full?: boolean;
}

export { IErrorMessage, ICheckBoxesProps };

const CheckBoxes: FunctionComponent<ICheckBoxesProps> = ({
  returnBoolean,
  onChange: onChangeCheckboxList,
  options,
  error,
  validationMessage,
  name,
  full,
  ...attrs
}): JSX.Element => {
  const formatedOptions = Array.isArray(options) ? options : [options];

  const [data, setData]: any = useState(formatedOptions);

  useEffect(() => setData(formatedOptions), [options]);

  const onChange = ({
    value,
  }: {
    event: React.MouseEvent<HTMLDivElement>;
    value: any;
  }): void => {
    if (value.disabled) {
      return;
    }

    const currentIndex = data.findIndex(
      ({ id }: { id: string }) => id === value.id
    );
    if (currentIndex < 0) {
      return;
    }

    data[currentIndex].checked = value.checked;
    setData([...data]);

    let currentValues = data.filter(
      ({ checked }: { checked: boolean }): boolean => checked
    );

    if (returnBoolean) {
      currentValues = !!currentValues.length;
    }

    onChangeCheckboxList({
      data,
      event: { target: { name } },
      value: currentValues,
    } as any);
  };

  return (
    <>
      {data.map(
        (
          {
            checked = false,
            disabled,
            id,
            label,
            name: checkboxName,
            required,
            show = true,
            value,
          }: any,
          key: number
        ): JSX.Element | null =>
          show ? (
            <Checkbox
              full={full}
              key={id || uuidv4()}
              onChange={onChange}
              checked={checked}
              disabled={disabled}
              id={id}
              label={label}
              name={checkboxName}
              required={required}
              show={show}
              value={value}
            />
          ) : null
      )}
      {error && <ErrorMessage message={validationMessage} />}
    </>
  );
};

CheckBoxes.defaultProps = {
  error: false,
  full: false,
  options: [],
  returnBoolean: false,
  validationMessage: "This field is required",
};

export default CheckBoxes;