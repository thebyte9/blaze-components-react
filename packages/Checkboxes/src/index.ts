import Checkboxes, { ICheckboxItem } from './Checkboxes';

import Checkbox from './Checkbox';

export { Checkbox, Checkboxes, ICheckboxItem };
