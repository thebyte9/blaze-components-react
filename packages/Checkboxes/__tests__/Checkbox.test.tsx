import { fireEvent, render } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import expect from "expect";
import React from "react";
import { Checkboxes } from "../src";

const options = [
  {
    id: 1,
    label: "first",
    value: "lorem ipsum",
  },
  {
    id: 2,
    label: "I accept",
    required: true,
    value: "accepted",
  },
  {
    id: 3,
    disabled: true,
    label: "Disabled",
    value: "",
  },
  {
    id: 4,
    label: "display none",
    show: false,
  },
];

const single = {
  label: "Single",
  value: "lorem ipsum",
};

const defaultProps = (override = {}) => ({
  onChange: () => ({}),
  options,
  ...override,
});

describe("Checkboxes component", () => {
  test("should be defined and renders correctly (snapshot)", () => {
    const wrapper = shallow(<Checkboxes {...defaultProps()} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test("should render and toggle single checkbox", () => {
    const override = {
      options: single,
      returnBoolean: true,
    };

    const wrapper = mount(<Checkboxes {...defaultProps(override)} />);

    expect(wrapper.find("input").length).toBe(1);

    expect(wrapper.find("input").at(0).prop("checked")).toBe(false);

    wrapper.find("input").at(0).simulate("click");

    expect(wrapper.find("input").at(0).prop("checked")).toBe(true);
  });

  test("should render multiple checkboxes and toggle", () => {
    let selectedBoxLabel: string = "";

    type TValue = object | any;

    const onChange = ({ value: [{ label }] }: TValue) => {
      selectedBoxLabel = label;
    };

    let override = {
      onChange,
    };

    const { rerender, getByTestId } = render(
      <Checkboxes {...defaultProps(override)} />
    );

    fireEvent.click(getByTestId("checkbox-1"));

    expect(selectedBoxLabel).toEqual("first");

    override = {
      ...override,
      ...{
        options: single,
      },
    };

    rerender(<Checkboxes {...defaultProps(override)} />);
  });

  test("can't interact when checkbox is disabled", () => {
    const wrapper = mount(<Checkboxes {...defaultProps()} />);

    wrapper.find(".form-field").at(2).simulate("click");

    expect(wrapper.find("input").at(2).prop("checked")).toBe(false);
  });
});
