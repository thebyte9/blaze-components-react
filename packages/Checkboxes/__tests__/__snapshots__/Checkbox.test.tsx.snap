// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes component should be defined and renders correctly (snapshot) 1`] = `
<Checkboxes
  onChange={[Function]}
  options={
    Array [
      Object {
        "id": 1,
        "label": "first",
        "value": "lorem ipsum",
      },
      Object {
        "id": 2,
        "label": "I accept",
        "required": true,
        "value": "accepted",
      },
      Object {
        "disabled": true,
        "id": 3,
        "label": "Disabled",
        "value": "",
      },
      Object {
        "id": 4,
        "label": "display none",
        "show": false,
      },
    ]
  }
  returnBoolean={false}
  utils={
    Object {
      "classNames": [Function],
      "uniqueId": [Function],
    }
  }
/>
`;
